server/test/performance/performance_test_annotations_backend.py

```diff
@@ -0,0 +1,215 @@
+        self.url_base = url_base
+        self.test_notes = self.create_info_dict()
+
+    def set_cell_count(self, dataset_name):
+        dataset_url = self.test_datasets[dataset_name]["dataset_url"]
+        headers = {"Content-Type": "application/octet-stream", "Cookie": cookie}
+        response = self.client.get(f"{self.url_base}{dataset_url}/api/v0.2/schema", headers=headers)
+        cell_count = json.loads(response._content)["schema"]["dataframe"]["nObs"]
+        self.test_datasets[dataset_name]["cell_count"] = cell_count
+
+    def create_info_dict(self):
+        request_info = {}
+        for dataset in self.test_datasets.keys():
+            request_info[dataset] = {}
+            for cat_count in self.annotations_category_count:
+                request_info[dataset][f"num_categories_{cat_count}"] = {}
+                for unique_labels in self.max_labels:
+                    request_info[dataset][f"num_categories_{cat_count}"][f"max_label_{unique_labels}"] = {}
+        return request_info
+
+    def create_annotations_dict_multi_process(self, dataset_name, category_count, label_max):
+        annotation_dict = {}
+        futures = []
+        categories = [f"Category{i}" for i in range(category_count)]
+        if not self.test_datasets[dataset_name]["num_cells"]:
+            self.set_cell_count(dataset_name)
+        with concurrent.futures.ProcessPoolExecutor(max_workers=5) as executor:
+            for category in categories:
+                futures.append(
+                    executor.submit(
+                        self.build_array_for_category,
+                        category,
+                        self.test_datasets[dataset_name]["num_cells"],
+                        label_max,
+                    )
+                )
+            for future in concurrent.futures.as_completed(futures):
+                try:
+                    result = future.result()
+                    category_name, cells = result
+                    annotation_dict[category_name] = pd.Series(cells, dtype="category")
+                except Exception as e:
+                    print(f"Issue creating the annotations dict: {e}")
+        return annotation_dict
+
+    def build_array_for_category(self, category_name, cell_count, label_max):
+        unique_label_count = label_max
+        labels = self.generate_labels(unique_label_count)
+        cells_per_label = int(cell_count / len(labels))
+        extra = cell_count % len(labels)
+        cells = []
+        for label in labels:
+            cells.extend([label] * cells_per_label)
+        cells.extend(["extra"] * extra)
+        rng = np.random.default_rng()
+        rng.shuffle(cells)
+        return category_name, cells
+
+    @staticmethod
+    def convert_to_fbs(annotation_dict):
+        df = pd.DataFrame(annotation_dict)
+        return encode_matrix_fbs(matrix=df, row_idx=None, col_idx=df.columns)
+
+    @staticmethod
+    def generate_labels(unique_label_count):
```

Comment:
Does this function produce a consistent number of labels? Line 126, makes me think the number of labels produces is variables. This could affect the reproducibility of the performance test.

Reply:
it generates a consistent number of labels, but the labels are of varying length

Reply:
got it. I had to parse the () closer.